Add `Observable.expectation` with support for measurement results and density matrices; Refactor `Collector` -> `Executor`

rmlarose opened the pull request:
Description
-----------

Part of #623.

Checklist
-----------

Check off the following once complete (or if not applicable) after opening the PR. The PR will be reviewed once this checklist is complete and all tests are passing.

- [x] I added unit tests for new code.
- [x] I used [type hints](https://www.python.org/dev/peps/pep-0484/) in function signatures.
- [x] I used [Google-style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) docstrings for functions.
- [x] I [updated the documentation](../blob/master/docs/CONTRIBUTING_DOCS.md) where relevant.

If some items remain, you can mark this a [draft pull request](https://github.blog/2019-02-14-introducing-draft-pull-requests/).

Tips
----

- If the validation check fails:

    1. Run `make check-types` (from the root directory of the repository) and fix any [mypy](https://mypy.readthedocs.io/en/stable/) errors.

    2. Run `make check-style` and fix any [flake8](http://flake8.pycqa.org) errors.

    3. Run `make format` to format your code with the [black](https://black.readthedocs.io/en/stable/index.html) autoformatter.

  For more information, check the [Mitiq style guidelines](https://mitiq.readthedocs.io/en/stable/contributing.html#style-guidelines).
  
- Write "Fixes #XYZ" in the description if this PR fixes Issue #XYZ.

andreamari commented on mitiq/observable/tests/test_observable.py:
```diff
@@ -32,6 +34,25 @@
+def execute(circuit: cirq.Circuit, shots: int = 8192) -> MeasurementResult:
+    result = cirq.Simulator().run(circuit, repetitions=shots)
+
+    n = sum(result._measurement_shape()[1].values())
+    bitstrings = [
+        list(map(int, np.binary_repr(value[0], width=n)))
+        for value in result.data.to_numpy()
+    ]
+    return MeasurementResult(
+        bitstrings,
+        qubit_indices=tuple(int(i) for i in result.data.columns[0].split(",")),
+    )
```
Not important for this test but, just to sketch the typical workflow that  the user should apply, we could find something simpler.

For example, I find `result.measurements` more intuitive with respect to `result.data`, since the output is very similar to the input of `MeasurementResult`.

The following should work, even in the case of multiple measurements:

```suggestion
    return MeasurementResult(
        result=np.column_stack(list(result.measurements.values())),
        qubit_indices=tuple(
            int(q) for k in result.measurements.keys() for q in k.split(",")
        ),
    )
```

rmlarose replied:
Thanks! Much simpler, went with this.

andreamari commented on mitiq/observable/observable.py:
```diff
@@ -94,16 +96,21 @@ def matrix(self, dtype: type = np.complex128) -> np.ndarray:
+            density_matrix = cast(np.ndarray, execute(circuit))
+            return np.trace(density_matrix @ self.matrix())
+        else:
+            raise NotImplementedError
```
Here we could rise a a more explicit error since this can happen often. 
For example: `"The output type of "execute" must be ... or ... but is  ....  ."`

E.g. the user may try the standard old-style executor that returns a float or may just forget annotation.

rmlarose replied:
Added.

rmlarose replied:
Note that support for batched versions, e.g. `List[mitiq.MeasurementResult]`, need to be added. Meant to add a TODO comment but forgot.

rmlarose commented:
Thanks @andreamari. In addition to the above comments I added a partial trace to the density matrix if the observable is on fewer qubits than the ansatz circuit, and added the density matrix executor to the last test which measures, e.g., <X1> on a three-qubit state.
